File: library/src/main/java/it/sephiroth/android/library/tooltip/TooltipTextDrawable.java
package it.sephiroth.android.library.tooltip;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.Outline;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PixelFormat;
import android.graphics.Point;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.support.annotation.Nullable;

class TooltipTextDrawable extends Drawable {
    public static final float ARROW_RATIO_DEFAULT = 1.4f;
    public static final float ALPHA_MAX = 255f;
    static final String TAG = "TooltipTextDrawable";
    private final RectF rectF;
    private final Path path;
    private final Point tmpPoint = new Point();
    private final Rect outlineRect = new Rect();
    private final Paint bgPaint;
    private final Paint stPaint;
    private final float arrowRatio;
    private final float ellipseSize;
    private Point point;
    private int padding = 0;
    private int arrowWeight = 0;
    private Tooltip.Gravity gravity;

    public TooltipTextDrawable(final Context context, final Tooltip.Builder builder) {

        TypedArray theme =
            context.getTheme().obtainStyledAttributes(null, R.styleable.TooltipLayout, builder.defStyleAttr, builder.defStyleRes);
        this.ellipseSize = theme.getDimensionPixelSize(R.styleable.TooltipLayout_ttlm_cornerRadius, 4);
        final int strokeWidth = theme.getDimensionPixelSize(R.styleable.TooltipLayout_ttlm_strokeWeight, 2);
        final int backgroundColor = theme.getColor(R.styleable.TooltipLayout_ttlm_backgroundColor, 0);
        final int strokeColor = theme.getColor(R.styleable.TooltipLayout_ttlm_strokeColor, 0);
        this.arrowRatio = theme.getFloat(R.styleable.TooltipLayout_ttlm_arrowRatio, ARROW_RATIO_DEFAULT);
        theme.recycle();

        this.rectF = new RectF();

        if (backgroundColor != 0) {
            bgPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
            bgPaint.setColor(backgroundColor);
            bgPaint.setStyle(Paint.Style.FILL);
        } else {
            bgPaint = null;
        }

        if (strokeColor != 0) {
            stPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
            stPaint.setColor(strokeColor);
            stPaint.setStyle(Paint.Style.STROKE);
            stPaint.setStrokeWidth(strokeWidth);
        } else {
            stPaint = null;
        }

        path = new Path();
    }

    @Override
    public void draw(final Canvas canvas) {
        if (null != bgPaint) {
            canvas.drawPath(path, bgPaint);
        }

        if (null != stPaint) {
            canvas.drawPath(path, stPaint);
        }
    }

    public void setAnchor(final Tooltip.Gravity gravity, int padding, @Nullable Point point) {
        if (gravity != this.gravity || padding != this.padding || !Utils.equals(this.point, point)) {
            this.gravity = gravity;
            this.padding = padding;
            this.arrowWeight = (int) ((float) padding / arrowRatio);

            if (null != point) {
                this.point = new Point(point);
            } else {
                this.point = null;
            }

            final Rect bounds = getBounds();
            if (!bounds.isEmpty()) {
                calculatePath(getBounds());
                invalidateSelf();
            }
        }
    }

    void calculatePath(Rect outBounds) {
        int left = outBounds.left + padding;
        int top = outBounds.top + padding;
        int right = outBounds.right - padding;
        int bottom = outBounds.bottom - padding;

        final float maxY = bottom - ellipseSize;
        final float maxX = right - ellipseSize;
        final float minY = top + ellipseSize;
        final float minX = left + ellipseSize;

        if (null != point && null != gravity) {
            calculatePathWithGravity(outBounds, left, top, right, bottom, maxY, maxX, minY, minX);
        } else {
            rectF.set(left, top, right, bottom);
            path.addRoundRect(rectF, ellipseSize, ellipseSize, Path.Direction.CW);
        }
    }

    private void calculatePathWithGravity(
        final Rect outBounds, final int left, final int top, final int right, final int bottom, final float maxY, final float maxX,
        final float minY, final float minX) {
        boolean drawPoint =
            isDrawPoint(left, top, right, bottom, maxY, maxX, minY, minX, tmpPoint, point, gravity, arrowWeight);
        clampPoint(left, top, right, bottom, tmpPoint);

        path.reset();

        // top/left
        path.moveTo(left + ellipseSize, top);

        if (drawPoint && gravity == Tooltip.Gravity.BOTTOM) {
            path.lineTo(left + tmpPoint.x - arrowWeight, top);
            path.lineTo(left + tmpPoint.x, outBounds.top);
            path.lineTo(left + tmpPoint.x + arrowWeight, top);
        }

        // top/right
        path.lineTo(right - ellipseSize, top);
        path.quadTo(right, top, right, top + ellipseSize);

        if (drawPoint && gravity == Tooltip.Gravity.LEFT) {
            path.lineTo(right, top + tmpPoint.y - arrowWeight);
            path.lineTo(outBounds.right, top + tmpPoint.y);
            path.lineTo(right, top + tmpPoint.y + arrowWeight);
        }

        // bottom/right
        path.lineTo(right, bottom - ellipseSize);
        path.quadTo(right, bottom, right - ellipseSize, bottom);

        if (drawPoint && gravity == Tooltip.Gravity.TOP) {
            path.lineTo(left + tmpPoint.x + arrowWeight, bottom);
            path.lineTo(left + tmpPoint.x, outBounds.bottom);
            path.lineTo(left + tmpPoint.x - arrowWeight, bottom);
        }

        // bottom/left
        path.lineTo(left + ellipseSize, bottom);
        path.quadTo(left, bottom, left, bottom - ellipseSize);

        if (drawPoint && gravity == Tooltip.Gravity.RIGHT) {
            path.lineTo(left, top + tmpPoint.y + arrowWeight);
            path.lineTo(outBounds.left, top + tmpPoint.y);
            path.lineTo(left, top + tmpPoint.y - arrowWeight);
        }

        // top/left
        path.lineTo(left, top + ellipseSize);
        path.quadTo(left, top, left + ellipseSize, top);
    }

    private static boolean isDrawPoint(
        final int left, final int top, final int right, final int bottom, final float maxY, final float maxX, final float minY,
        final float minX, final Point tmpPoint, final Point point, final Tooltip.Gravity gravity,
        final int arrowWeight) {
        boolean drawPoint = false;
        tmpPoint.set(point.x, point.y);

        if (gravity == Tooltip.Gravity.RIGHT || gravity == Tooltip.Gravity.LEFT) {
            if (tmpPoint.y >= top && tmpPoint.y <= bottom) {
                if (top + tmpPoint.y + arrowWeight > maxY) {
                    tmpPoint.y = (int) (maxY - arrowWeight - top);
                } else if (top + tmpPoint.y - arrowWeight < minY) {
                    tmpPoint.y = (int) (minY + arrowWeight - top);
                }
                drawPoint = true;
            }
        } else {
            if (tmpPoint.x >= left && tmpPoint.x <= right) {
                if (tmpPoint.x >= left && tmpPoint.x <= right) {
                    if (left + tmpPoint.x + arrowWeight > maxX) {
                        tmpPoint.x = (int) (maxX - arrowWeight - left);
                    } else if (left + tmpPoint.x - arrowWeight < minX) {
                        tmpPoint.x = (int) (minX + arrowWeight - left);
                    }
                    drawPoint = true;
                }
            }
        }
        return drawPoint;
    }

    private static void clampPoint(
        final int left, final int top, final int right, final int bottom, final Point tmpPoint) {
        if (tmpPoint.y < top) {
            tmpPoint.y = top;
        } else if (tmpPoint.y > bottom) {
            tmpPoint.y = bottom;
        }
        if (tmpPoint.x < left) {
            tmpPoint.x = left;
        }
        if (tmpPoint.x > right) {
            tmpPoint.x = right;
        }
    }

    @Override
    protected void onBoundsChange(final Rect bounds) {
        super.onBoundsChange(bounds);
        calculatePath(bounds);
    }

    public float getRadius() {
        return ellipseSize;
    }

    @Override
    public int getAlpha() {
        return bgPaint.getAlpha();
    }

    @Override
    public void setAlpha(final int alpha) {
        bgPaint.setAlpha(alpha);
        stPaint.setAlpha(alpha);
    }

    @Override
    public void setColorFilter(final ColorFilter cf) {
    }

    @Override
    public int getOpacity() {
        return PixelFormat.TRANSLUCENT;
    }

    @TargetApi (Build.VERSION_CODES.LOLLIPOP)
    @Override
    public void getOutline(Outline outline) {
        copyBounds(outlineRect);
        outlineRect.inset(padding, padding);
        outline.setRoundRect(outlineRect, getRadius());
        if (getAlpha() < 255) {
            outline.setAlpha(0);
        }
        //outline.setAlpha(getAlpha() / ALPHA_MAX);
    }
}

File: neofecttooltip/src/main/java/it/sephiroth/android/library/xtooltip/TooltipTextDrawable.kt
package it.sephiroth.android.library.xtooltip


import android.annotation.TargetApi
import android.content.Context
import android.graphics.*
import android.graphics.drawable.Drawable
import android.os.Build
import android.support.v4.util.ObjectsCompat
import it.sephiroth.android.library.tooltip.R
//import androidx.core.util.ObjectsCompat
import timber.log.Timber

/**
 * Created by alessandro crugnola on 12/12/15.
 * alessandro.crugnola@gmail.com
 *
 *
 * LICENSE
 * Copyright 2015 Alessandro Crugnola
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy,
 * modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software
 * is furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT
 * OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
internal class TooltipTextDrawable(context: Context, builder: XTooltip.Builder) : Drawable() {
    private val rectF: RectF
    private val path: Path
    private val tmpPoint = Point()
    private val outlineRect = Rect()
    private val bgPaint: Paint?
    private val stPaint: Paint?
    private val arrowRatio: Float
    private val radius: Float
    private var point: Point? = null
    private var padding = 0
    private var arrowWeight = 0
    private var gravity: XTooltip.Gravity? = null

    init {

        val theme = context.theme.obtainStyledAttributes(
                null,
                R.styleable.TooltipLayout,
                builder.defStyleAttr,
                builder.defStyleRes
                                                        )
        this.radius = theme.getDimensionPixelSize(R.styleable.TooltipLayout_ttlm_cornerRadius, 4).toFloat()
        val strokeWidth = theme.getDimensionPixelSize(R.styleable.TooltipLayout_ttlm_strokeWeight, 2)
        val backgroundColor = theme.getColor(R.styleable.TooltipLayout_ttlm_backgroundColor, 0)
        val strokeColor = theme.getColor(R.styleable.TooltipLayout_ttlm_strokeColor, 0)
        this.arrowRatio = theme.getFloat(R.styleable.TooltipLayout_ttlm_arrowRatio, ARROW_RATIO_DEFAULT)
        theme.recycle()

        this.rectF = RectF()

        if (backgroundColor != 0) {
            bgPaint = Paint(Paint.ANTI_ALIAS_FLAG)
            bgPaint.color = backgroundColor
            bgPaint.style = Paint.Style.FILL
        } else {
            bgPaint = null
        }

        if (strokeColor != 0) {
            stPaint = Paint(Paint.ANTI_ALIAS_FLAG)
            stPaint.color = strokeColor
            stPaint.style = Paint.Style.STROKE
            stPaint.strokeWidth = strokeWidth.toFloat()
        } else {
            stPaint = null
        }

        path = Path()
    }

    override fun draw(canvas: Canvas) {
        bgPaint?.let {
            canvas.drawPath(path, it)
        }

        stPaint?.let {
            canvas.drawPath(path, it)
        }
    }

    fun setAnchor(gravity: XTooltip.Gravity, padding: Int, point: Point?) {
        if (gravity !== this.gravity || padding != this.padding || !ObjectsCompat.equals(this.point, point)) {
            this.gravity = gravity
            this.padding = padding
            this.arrowWeight = (padding.toFloat() / arrowRatio).toInt()

            point?.let {
                this.point = Point(it)
            } ?: run {
                this.point = null
            }

            if (!bounds.isEmpty) {
                calculatePath(bounds)
                invalidateSelf()
            }
        }
    }

    private fun calculatePath(outBounds: Rect) {
        val left = outBounds.left + padding
        val top = outBounds.top + padding
        val right = outBounds.right - padding
        val bottom = outBounds.bottom - padding

        val maxY = bottom - radius
        val maxX = right - radius
        val minY = top + radius
        val minX = left + radius

        if (null != point && null != gravity) {
            calculatePathWithGravity(outBounds, left, top, right, bottom, maxY, maxX, minY, minX)
        } else {
            rectF.set(left.toFloat(), top.toFloat(), right.toFloat(), bottom.toFloat())
            path.addRoundRect(rectF, radius, radius, Path.Direction.CW)
        }
    }

    private fun calculatePathWithGravity(
            outBounds: Rect,
            left: Int,
            top: Int,
            right: Int,
            bottom: Int,
            maxY: Float,
            maxX: Float,
            minY: Float,
            minX: Float
                                        ) {
        val drawPoint =
                isDrawPoint(left, top, right, bottom, maxY, maxX, minY, minX, tmpPoint, point!!, gravity, arrowWeight)

        Timber.v("drawPoint: $drawPoint")

        clampPoint(left, top, right, bottom, tmpPoint)

        path.reset()

        // top/left
        path.moveTo(left + radius, top.toFloat())

        if (drawPoint && gravity === XTooltip.Gravity.BOTTOM) {
            path.lineTo((left + tmpPoint.x - arrowWeight).toFloat(), top.toFloat())
            path.lineTo((left + tmpPoint.x).toFloat(), outBounds.top.toFloat())
            path.lineTo((left + tmpPoint.x + arrowWeight).toFloat(), top.toFloat())
        }

        // top/right
        path.lineTo(right - radius, top.toFloat())
        path.quadTo(right.toFloat(), top.toFloat(), right.toFloat(), top + radius)

        if (drawPoint && gravity === XTooltip.Gravity.LEFT) {
            path.lineTo(right.toFloat(), (top + tmpPoint.y - arrowWeight).toFloat())
            path.lineTo(outBounds.right.toFloat(), (top + tmpPoint.y).toFloat())
            path.lineTo(right.toFloat(), (top + tmpPoint.y + arrowWeight).toFloat())
        }

        // bottom/right
        path.lineTo(right.toFloat(), bottom - radius)
        path.quadTo(right.toFloat(), bottom.toFloat(), right - radius, bottom.toFloat())

        if (drawPoint && gravity === XTooltip.Gravity.TOP) {
            path.lineTo((left + tmpPoint.x + arrowWeight).toFloat(), bottom.toFloat())
            path.lineTo((left + tmpPoint.x).toFloat(), outBounds.bottom.toFloat())
            path.lineTo((left + tmpPoint.x - arrowWeight).toFloat(), bottom.toFloat())
        }

        // bottom/left
        path.lineTo(left + radius, bottom.toFloat())
        path.quadTo(left.toFloat(), bottom.toFloat(), left.toFloat(), bottom - radius)

        if (drawPoint && gravity === XTooltip.Gravity.RIGHT) {
            path.lineTo(left.toFloat(), (top + tmpPoint.y + arrowWeight).toFloat())
            path.lineTo(outBounds.left.toFloat(), (top + tmpPoint.y).toFloat())
            path.lineTo(left.toFloat(), (top + tmpPoint.y - arrowWeight).toFloat())
        }

        // top/left
        path.lineTo(left.toFloat(), top + radius)
        path.quadTo(left.toFloat(), top.toFloat(), left + radius, top.toFloat())
    }

    override fun onBoundsChange(bounds: Rect) {
        super.onBoundsChange(bounds)
        calculatePath(bounds)
    }

    override fun getAlpha(): Int {
        return bgPaint?.alpha ?: run { 0 }
    }

    override fun setAlpha(alpha: Int) {
        bgPaint?.alpha = alpha
        stPaint?.alpha = alpha
    }

    override fun setColorFilter(cf: ColorFilter?) {}

    override fun getOpacity(): Int {
        return PixelFormat.TRANSLUCENT
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    override fun getOutline(outline: Outline) {
        copyBounds(outlineRect)
        outlineRect.inset(padding, padding)
        outline.setRoundRect(outlineRect, radius)
        if (alpha < 255) {
            outline.alpha = 0f
        }
        //outline.setAlpha(getAlpha() / ALPHA_MAX);
    }

    companion object {
        const val ARROW_RATIO_DEFAULT = 1.4f

        private fun isDrawPoint(
                left: Int, top: Int, right: Int, bottom: Int, maxY: Float, maxX: Float, minY: Float,
                minX: Float, tmpPoint: Point, point: Point, gravity: XTooltip.Gravity?,
                arrowWeight: Int
                               ): Boolean {
            Timber.i("isDrawPoint: $left, $top, $right, $bottom, $maxX, $maxY, $minX, $minY, $point, $arrowWeight")
            var drawPoint = false
            tmpPoint.set(point.x, point.y)

            if (gravity === XTooltip.Gravity.RIGHT || gravity === XTooltip.Gravity.LEFT) {
                if (tmpPoint.y in top..bottom) {
                    if (top + tmpPoint.y + arrowWeight > maxY) {
                        tmpPoint.y = (maxY - arrowWeight.toFloat() - top.toFloat()).toInt()
                    } else if (top + tmpPoint.y - arrowWeight < minY) {
                        tmpPoint.y = (minY + arrowWeight - top).toInt()
                    }
                    drawPoint = true
                }
            } else {
                if (tmpPoint.x in left..right) {
                    if (tmpPoint.x in left..right) {
                        if (left + tmpPoint.x + arrowWeight > maxX) {
                            tmpPoint.x = (maxX - arrowWeight.toFloat() - left.toFloat()).toInt()
                        } else if (left + tmpPoint.x - arrowWeight < minX) {
                            tmpPoint.x = (minX + arrowWeight - left).toInt()
                        }
                        drawPoint = true
                    }
                }
            }
            return drawPoint
        }

        private fun clampPoint(left: Int, top: Int, right: Int, bottom: Int, tmpPoint: Point) {
            if (tmpPoint.y < top) {
                tmpPoint.y = top
            } else if (tmpPoint.y > bottom) {
                tmpPoint.y = bottom
            }
            if (tmpPoint.x < left) {
                tmpPoint.x = left
            }
            if (tmpPoint.x > right) {
                tmpPoint.x = right
            }
        }
    }
}
